src/server/pfs/server/driver_new_storage.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"golang.org/x/net/context"
+)
+
+const (
+	compacted = "compacted"
+)
+
+func init() {
+	// This is kind of weird, but I think it might be necessary.
+	// Since we build our protos with gogoproto, we need to register
+	// the types that can be embedded in an Any type with the gogoproto
+	// proto package because building the protos just registers them with the
+	// normal proto package.
+	proto.RegisterType((*pfs.Merge)(nil), "pfs.Merge")
+	proto.RegisterType((*pfs.Shard)(nil), "pfs.Shard")
+}
+
+var (
+	// (bryce) need to expose as configuration.
+	memThreshold int64 = 1024 * chunk.MB
+)
+
+func (d *driver) startCommitNewStorageLayer(txnCtx *txnenv.TransactionContext, id string, parent *pfs.Commit, branch string, provenance []*pfs.CommitProvenance, description string) (*pfs.Commit, error) {
+	commit, err := d.startCommit(txnCtx, id, parent, branch, provenance, description)
+	if err != nil {
+		return nil, err
+	}
+	d.fs = d.storage.New(context.Background(), commit.ID, fileset.WithMemThreshold(memThreshold))
+	return commit, nil
+}
+
+func (d *driver) finishCommitNewStorageLayer(txnCtx *txnenv.TransactionContext, commit *pfs.Commit, description string) (retErr error) {
+	if err := d.checkIsAuthorizedInTransaction(txnCtx, commit.Repo, auth.Scope_WRITER); err != nil {
+		return err
+	}
+	commitInfo, err := d.resolveCommit(txnCtx.Stm, commit)
+	if err != nil {
+		return err
+	}
+	if commitInfo.Finished != nil {
+		return pfsserver.ErrCommitFinished{commit}
+	}
+	if description != "" {
+		commitInfo.Description = description
+	}
+	// Close in-memory file set (serializes in-memory part).
+	if err := d.fs.Close(); err != nil {
+		return err
+	}
+	defer func() {
+		if retErr == nil {
+			d.fs = nil
+		}
+	}()
+	// Merge the commit with its parent.
+	// (bryce) this should be where the level based compaction scheme is used.
+	// right now we are eagerly compacting at every commit.
+	fileSets := []string{commitInfo.Commit.ID}
+	parentCommit := commitInfo.ParentCommit
+	if parentCommit != nil {
+		// (bryce) how to handle parent commit that is not closed?
+		parentCommitInfo, err := d.resolveCommit(txnCtx.Stm, parentCommit)
+		if err != nil {
+			return err
+		}
+		fileSets = append(fileSets, path.Join(parentCommitInfo.Commit.ID, compacted))
+	}
+	if err := d.merge(context.Background(), path.Join(commitInfo.Commit.ID, compacted), fileSets); err != nil {
+		return err
+	}
+	// (bryce) need size.
+	commitInfo.SizeBytes = uint64(0)
+	commitInfo.Finished = now()
+	return d.writeFinishedCommit(txnCtx.Stm, commit, commitInfo)
+}
+
+// (bryce) should have a context going through here.
+func (d *driver) putFilesNewStorageLayer(pachClient *client.APIClient, server *putFileServer) error {
+	// (bryce) how to handle no commit started?
+	for req, err := server.Recv(); err == nil; req, err = server.Recv() {
+		if err := d.fs.WriteHeader(&tar.Header{Name: req.File.Path}); err != nil {
```

Comment:
This looks like it's not going to handle large files correctly, because they'll have requests with just data and now `File` set.

Reply:
Yeah, I just wanted to get a basic integration working, I can work on this a bit more or just add another note so we can get this reviewed and merged? There is still a decent bit of integration work to do and I think it would be best to do it in pieces rather than have another put object improvements.

Reply:
That seems fine, I would just sanity check that there's a test that fails when this code gets enabled so we don't accidentally forget it.

Reply:
Made a note.